Feature/update contracts sdk with snickerdoodle wallet

seansing9 opened the pull request:
### Release Notes
[JIRA Link]()

#### Summary:
This PR updates the contracts-sdk to new functionalities of the SnickerdoodleWallet and Factory contracts.

seansing committed:
WIP update snickerdoodlewallet in contracts sdk

seansing committed:
Merge branch 'feature/new-contracts-package' into feature/update-contracts-sdk-with-snickerdoodle-wallet

seansing committed:
Merge branch 'feature/new-contracts-package' into feature/update-contracts-sdk-with-snickerdoodle-wallet

seansing committed:
Merge branch 'feature/new-contracts-package' into feature/update-contracts-sdk-with-snickerdoodle-wallet

seansing committed:
Merge branch 'feature/new-contracts-package' into feature/update-contracts-sdk-with-snickerdoodle-wallet

seansing committed:
Merge branch 'feature/new-contracts-package' into feature/update-contracts-sdk-with-snickerdoodle-wallet

seansing committed:
complete contracts-sdk update for Snickerdoodle wallet and factory

seansing9 committed:
Merge branch 'feature/new-contracts-package' into feature/update-contracts-sdk-with-snickerdoodle-wallet

seansing9 committed:
update objects

seansing9 committed:
WIP update contracts-sdk and crypto utils

seansing9 committed:
Merge branch 'feature/new-contracts-package' into feature/update-contracts-sdk-with-snickerdoodle-wallet

seansing9 committed:
update contracts-sdk

seansing9 committed:
Merge branch 'feature/new-contracts-package' into feature/update-contracts-sdk-with-snickerdoodle-wallet

seansing9 committed:
updated signature components

seansing9 committed:
updated cryptoutils, included unit test

seansing9 committed:
update function name

seansing9 committed:
clean up

SnickerChar commented on packages/node-utils/src/implementations/CryptoUtils.ts:
```diff
@@ -32,6 +32,14 @@ import {
   ED25519PrivateKey,
+  P256PublicKeyComponent,
+  PasskeyId,
+  PasskeyPublicKeyPointX,
```
Should this be P256PublicKeyPointX rather than passkey, keeping it consistent with the P256Signature stuff?

SnickerChar commented on packages/node-utils/src/implementations/CryptoUtils.ts:
```diff
@@ -661,4 +672,167 @@ export class CryptoUtils implements ICryptoUtils {
+
+  public parseRawP256PublicKey(
+    id: PasskeyId,
+    publicKeyArray: Uint8Array,
```
Right now we have a PublicKey defined and used as a binary hex string; it might be better to make this take that object. A new P256PublicKey primitive would be really good. Also, for PasskeyId, is that really required? It's part of P256PublicKeyComponent but it's not used by this method; I think we could exclude it safely.

SnickerChar commented on packages/node-utils/src/implementations/CryptoUtils.ts:
```diff
@@ -661,4 +672,167 @@ export class CryptoUtils implements ICryptoUtils {
+    // Third value tells you the type of the next value which MUST be an integer (0x02) if this is a signature array
+    const metadataIndicatorByte = pubKeyView[2];
+    // Third byte MUST be equal to 48 if this is a legitimate public key array
+    console.assert(
```
We should make this method return a Result<> so that this error could be returned. Doesn't need to be a ResultAsync<>, but this error should be properly handled

SnickerChar commented on packages/objects/src/businessObjects/AuthenticatorData.ts:
Empty file

seansing9 committed:
fix from comments

SnickerChar commented on packages/node-utils/src/implementations/CryptoUtils.ts:
```diff
@@ -661,4 +673,175 @@ export class CryptoUtils implements ICryptoUtils {
+
+  // returns a 64-byte ArrayBuffer containing r and s concatenated together
+  public parseRawP256Signature(
+    signatureArray: ArrayBuffer,
```
Let's change the name and type to signature: P256Signature

SnickerChar commented on packages/node-utils/src/implementations/CryptoUtils.ts:
```diff
@@ -661,4 +673,175 @@ export class CryptoUtils implements ICryptoUtils {
+  // returns a 64-byte ArrayBuffer containing r and s concatenated together
+  public parseRawP256Signature(
+    signatureArray: ArrayBuffer,
+    msgPayload: string,
```
Let's make this take clientDataJson: JSONString, authenticatorData: AuthenticatorData (string), and follow Todd's example on line 328

SnickerChar commented on packages/node-utils/src/implementations/CryptoUtils.ts:
```diff
@@ -661,4 +673,175 @@ export class CryptoUtils implements ICryptoUtils {
+  }
+
+  // returns a 64-byte ArrayBuffer containing r and s concatenated together
+  public parseRawP256Signature(
```
Turns out this method isn't needed, I have this already baked into SimpleWebAuthn library.

SnickerChar commented on packages/contracts-sdk/src/interfaces/objects/AuthenticatorData.ts:
```diff
@@ -0,0 +1,10 @@
+import { JSONString } from "@snickerdoodlelabs/objects";
+import { BytesLike } from "ethers";
+
+export class AuthenticatorData {
```
Need a method in CryptoUtils parseClientDataJson(clientDataJSON: JSONString): Result<AuthenticatorData, whatever>

SnickerChar commented on packages/objects/src/businessObjects/P256SignatureComponent.ts:
```diff
@@ -0,0 +1,5 @@
+import { P256SignatureR, P256SignatureS } from "@objects/primitives/index.js";
+
+export class P256SignatureComponent {
```
Maybe ParsedP256Signature, or just make it P256SignatureComponents

SnickerChar commented on packages/objects/src/businessObjects/P256PublicKeyComponent.ts:
```diff
@@ -0,0 +1,13 @@
+  P256PublicKeyPointY,
+} from "@objects/primitives/index.js";
+
+export class P256PublicKeyComponent {
```
Name -=> P256PublicKeyComponents, drop the keyId

SnickerChar commented on packages/node-utils/src/interfaces/ICryptoUtils.ts:
```diff
@@ -159,7 +165,22 @@ export interface ICryptoUtils {
+  parseRawP256Signature(
+    signatureArray,
+    msgPayload,
+  ): P256SignatureComponentArrayBuffer;
```
Make this return a P256SignatureComponents object, can be passed to the contract layer, make sure it takes strings not arrays

SnickerChar commented on packages/node-utils/src/interfaces/ICryptoUtils.ts:
```diff
@@ -159,7 +165,22 @@ export interface ICryptoUtils {
+  parseRawP256PublicKey(
+    publicKey: P256PublicKey,
+  ): Result<
+    { x: P256PublicKeyPointX; y: P256PublicKeyPointY },
```
Return P256PublicKeyComponents (minus the key ID)

SnickerChar commented on packages/contracts-sdk/src/interfaces/ISnickerdoodleWalletContract.ts:
```diff
@@ -0,0 +1,79 @@
+  >;
+
+  addP256KeyWithP256Key(
+    keyId: PasskeyId,
```
PasskeyId -> WebauthnCredentialId

SnickerChar reviewed (commented):
Notes after discussion

seansing9 committed:
fix from discussion

SnickerChar commented on packages/contracts-sdk/src/interfaces/ISnickerdoodleWalletContract.ts:
```diff
@@ -0,0 +1,86 @@
+
+  addP256KeyWithP256Key(
+    keyId: WebauthnCredentialId,
+    authenticatorData: string,
```
Primitive type string AuthenticatorData

seansing9 committed:
rename AuthenticatorData to P256VerificationData

SnickerChar committed:
Fixed methods for generating challenges.
Updated package version for publishing

SnickerChar committed:
Publish node-utils

seansing9 committed:
implement generate challenge

seansing9 committed:
Merge branch 'feature/new-contracts-package' into feature/update-contracts-sdk-with-snickerdoodle-wallet

seansing9 committed:
update contracts sdk with latest functions

seansing9 committed:
Merge branch 'feature/new-contracts-package' into feature/update-contracts-sdk-with-snickerdoodle-wallet

SnickerChar committed:
Change challenge methods to static

seansing9 committed:
update generate challenge functions, add @hexagon/base64, add unit test

seansing9 committed:
update version, fix interface

SnickerChar commented on packages/contracts-sdk/src/implementations/SnickerdoodleWalletContract.ts:
```diff
@@ -281,4 +312,11 @@ export class SnickerdoodleWalletContract
     return new SnickerdoodleWalletContractError(msg, e, transaction);
   }
+
+  private isoBase64fromBuffer(
```
This method isn't needed now, should remove it, along with the import

SnickerChar commented on packages/contracts-sdk/src/implementations/SnickerdoodleWalletContract.ts:
```diff
@@ -132,8 +162,9 @@ export class SnickerdoodleWalletContract
-  ): Result<string, InvalidParametersError> {
-    return ok(evmAccountAddress);
+  ): Result<Uint8Array, InvalidParametersError> {
+    // Remove 0x, convert to Uint8Array
```
You should check this out, and make sure it works with both 0x123 and just 123. I think that .encode will handle both versions, you should test that.

seansing9 committed:
normalize evm address before generating challenge

SnickerChar committed:
Added contract addresses to chain information.
Made Base and Base Sepolia control chains

SnickerChar committed:
Change ChainInformation because the idea of different info only for the control chain no longer makes sense.
All chains may have the contract stack (indicated by contractStackDeployed).
The control chain is configured already in the IP but doesn't need different data

SnickerChar committed:
Removed unnecessary msgPayload parameter from parseRawP256Signature

seansing9 committed:
Merge branch 'feature/new-contracts-package' into feature/update-contracts-sdk-with-snickerdoodle-wallet

seansing9 committed:
updated contracts sdk functions and abi

seansing9 committed:
include operator gateway proxy deployment, updated dockerfiles, removed viem

seansing9 committed:
merged feature/new-contracts-package

seansing9 committed:
updated addresses in chains config, add scripts to contracts-sdk package json

SnickerChar committed:
Fix non-export of BlockchainErrorMapper

SnickerChar committed:
Merge branch 'feature/update-contracts-sdk-with-snickerdoodle-wallet' of https://github.com/SnickerdoodleLabs/protocol into feature/update-contracts-sdk-with-snickerdoodle-wallet

seansing9 committed:
update contracts sdk

seansing9 committed:
add asn1 prefix to parsing p256 public key
